storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp

```diff
@@ -0,0 +1,405 @@
+  }
+
+  {
+    std::lock_guard<std::mutex> guard(connectionMutex);
+    require(ndbConnection == nullptr);
+
+    int retCode   = 0;
+    ndbConnection = new Ndb_cluster_connection(connection_string, node_ids[0]);
+    retCode       = ndbConnection->connect(connection_retries, connection_retry_delay_in_sec, 0);
+    if (retCode != 0) {
+      return RS_SERVER_ERROR(ERROR_002 + std::string(" RetCode: ") + std::to_string(retCode));
+    }
+
+    retCode = ndbConnection->wait_until_ready(30, 0);
+    if (retCode != 0) {
+      return RS_SERVER_ERROR(ERROR_003 + std::string(" RetCode: ") + std::to_string(retCode) +
+                             std::string(" Lastest Error: ") +
+                             std::to_string(__instance->ndbConnection->get_latest_error()) +
+                             std::string(" Lastest Error Msg: ") +
+                             std::string(__instance->ndbConnection->get_latest_error_msg()));
+    }
+  }
+
+  {
+    std::lock_guard<std::mutex> guardInfo(connectionInfoMutex);
+    stats.connection_state = CONNECTED;
+  }
+
+  LOG_INFO("RonDB connection and object pool initialized");
+  return RS_OK;
+}
+
+//--------------------------------------------------------------------------------------------------
+
+RS_Status RDRSRonDBConnection::GetInstance(RDRSRonDBConnection **rdrsRonDBConnection) {
+  if (__instance == nullptr) {
+    LOG_ERROR(ERROR_035);
+    return RS_SERVER_ERROR(ERROR_035);
+  }
+
+  std::lock_guard<std::mutex> guardInfo(__instance->connectionInfoMutex);
+  if (__instance->ndbConnection == nullptr || __instance->stats.connection_state != CONNECTED ||
+      __instance->stats.is_shutdown) {
+    LOG_ERROR(ERROR_033);
+    return RS_SERVER_ERROR(ERROR_033);
+  }
+
+  *rdrsRonDBConnection = __instance;
+  return RS_OK;
+}
+
+//--------------------------------------------------------------------------------------------------
+
+RS_Status RDRSRonDBConnection::GetNdbObject(Ndb **ndb_object) {
+
+  {
+    bool is_shutdown;
+    bool reconnection_in_progress;
+    STATE connection_state = DISCONNECTED;
+    {
+      std::lock_guard<std::mutex> guardInfo(connectionInfoMutex);
+      is_shutdown              = stats.is_shutdown;
+      reconnection_in_progress = stats.is_reconnection_in_progress;
+      connection_state         = stats.connection_state;
+    }
+
+    if (is_shutdown) {
+      LOG_ERROR(ERROR_037);
+      return RS_SERVER_ERROR(ERROR_037);
+    }
+
+    if (connection_state != CONNECTED) {
+      if (!reconnection_in_progress) {
+        // If previous reconnection attempts have failed then
+        // restart the reconnection process
+        LOG_DEBUG("GetNdbObject triggered reconnection");
+        Reconnect();
+      }
+
+      LOG_ERROR(ERROR_033 + std::string(" Connection State: ") + std::to_string(connection_state) +
+                std::string(" Reconnection State: ") + std::to_string(reconnection_in_progress));
+      return RS_SERVER_ERROR(ERROR_033);
+    }
+  }
+
```

Comment:
What is the purpose of releasing the mutex on connectionInfoMutex to immediately seize it again?

Reply:
the first time it had only one lock and the next time it takes two locks